cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/GitHubHttpHandler.java

```diff
@@ -0,0 +1,126 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.gateway.handlers;
+
+import com.google.gson.Gson;
+import com.google.inject.Inject;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.gateway.handlers.util.AbstractAppFabricHttpHandler;
+import io.cdap.cdap.internal.app.store.GitHubStore;
+import io.cdap.cdap.internal.github.GitHubRepo;
+import io.cdap.http.HttpResponder;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponseStatus;
+
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+
+@Path(Constants.Gateway.API_VERSION_3)
+public class GitHubHttpHandler extends AbstractAppFabricHttpHandler {
+  private static final Gson GSON = new Gson();
+  private final GitHubStore gitStore;
+
+
+  @Inject
+  GitHubHttpHandler(GitHubStore gitStore) {
+    this.gitStore = gitStore;
+  }
+
+  @Path("repos/githt")
+  @GET
+  public void getRepos(HttpRequest request, HttpResponder responder) throws Exception {
+    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(gitStore.getRepos()));
+  }
+
+  /**
+   * Returns user repository information
+   */
+  @Path("repos/github/{repo}")
+  @GET
+  public void getRepoInfo(HttpRequest request, HttpResponder responder,
+                          @PathParam("repo") String repo) throws Exception {
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(gitStore.getRepo(repo)));
+  }
+
+  @Path("repos/github/{repo}")
+  @PUT
+  public void addRepoInfo(FullHttpRequest request, HttpResponder responder,
+                          @PathParam("repo") String repo) throws Exception {
+
+    GitHubRepo githubRequest = GSON.fromJson(request.content().toString(StandardCharsets.UTF_8)
+        , GitHubRepo.class);
+
+    String htmlUrl =  parseUrl(githubRequest.getUrl());
+
+    String authString  = "invalid";
+    if (githubRequest.getAuthMethod().equals("username and password")) {
+      authString = "Basic " + Base64.getEncoder().encodeToString(
+            (githubRequest.getUsername() + ":" + githubRequest.getPassword()).getBytes());
+    } else if (githubRequest.getAuthMethod().equals("authorization token")) {
+      authString = "token " + githubRequest.getAuthToken();
+    }
+    //test that the GitHub repository information is valid
+    if (checkGitConnection(htmlUrl, authString)) {
```

Comment:
It is better to provide a `Test connection` end point that can test without saving.

Reply:
Created endpoint "repos/github/{repo}/testconnection" for this.